packages/agents/src/agents/Chameleon/ContextualizedChat.ts

```diff
@@ -0,0 +1,294 @@
+    const results = await this._collections[type].search(context, -1);
+
+    // Aggregate as many as possible
+    const msgs: {
+      msg: ChatMessage;
+      chunkMetadataPtr: ChunkMetadataPtr;
+    }[] = [];
+    let tokenCounter = 0;
+
+    const addMsg = (chunkText: string, metadata: DocumentMetadata): boolean => {
+      if (tokenCounter + metadata.tokens > tokenLimit) {
+        return false;
+      }
+
+      const chunkMetadataPtr = metadata.chunkMetadataPtr;
+
+      const msg = JSON.parse(chunkText) as ChatMessage;
+      msgs.push({ msg, chunkMetadataPtr });
+      tokenCounter += metadata.tokens;
+      return true;
+    }
+
+    for (const result of results) {
+      const chunkText = result.text();
+      const chunkMetadata = result.metadata();
+
+      if (!chunkMetadata) {
+        throw Error("chunkmetadata is missing, this should never happen")
+      }
+
+      if (!addMsg(chunkText, chunkMetadata)) {
+        break;
+      }
+    }
+
+    return {
+      msgs,
+      tokens: tokenCounter
+    };
+  }
+
+  private async _processNewMessages(): Promise<void> {
+    // ensure all new messages have been processed
+    await this._processNewMessagesByType("persistent");
+    await this._processNewMessagesByType("temporary");
+  }
+
+  private async _processNewMessagesByType(type: ChatLogType): Promise<void> {
+    const messages = this._chat.chatLogs.get(type).msgs;
+
+    // If no messages exist
+    if (messages.length === 0) {
+      return Promise.resolve();
+    }
+
+    const lastProcessedIdx = this._getLastProcessedMessageIndex(type);
+
+    // If we've already processed all messages
+    if (lastProcessedIdx === messages.length - 1) {
+      return Promise.resolve();
+    }
+
+    // Process all messages from lastProcessedIdx -> messages.length
+    for (let i = lastProcessedIdx + 1; i < messages.length; ++i) {
+      const message = messages[i];
+      await this._processNewMessage(message, i, type);
+    }
+  }
+
+  private async _processNewMessage(message: ChatMessage, msgPtr: number, type: ChatLogType): Promise<void> {
+
+    const chunks: string[] = [];
+
+    // If the message is large and requires chunking
+    if (this._chunker.shouldChunk(message)) { 
+      chunks.push(...this._chunker.chunk(message));
+    } else {
+      // TODO: should this be what is used for embedding?
+      //       should we instead just extract the values of the struct?
+      chunks.push(JSON.stringify(message));
+    }
+
+    const chunkMetadata = this._chunkMetadata[type];
+
+    // Save the starting chunk index (used later for document index)
+    const startChunkIdx = chunkMetadata.length;
+
+    // Add message index pointers for each new chunk
+    for (const _ of chunks) {
+      chunkMetadata.push({
+        msgPtr
+      });
+    }
+
+    // If we're processing persistent messages,
+    // and it isn't a large message, early out
+    if (type === "persistent" && chunks.length === 1) {
```

Comment:
If we need to disambiguate per type, and pass `type` around, is it still worth to have chunkMetadata have types in a single struct?

Reply:
Talked about the trade-off of this, would create more code as we'd need to support divergent paths.